### DISCUSSION TITLE
input or output?

### DISCUSSION THREAD

**Discussion by zzzh** [zephyrproject-rtos/zephyr#37490] [Q&A] [answered]
/** GPIO pin is active (has logical value '1') in low state. */
#define GPIO_ACTIVE_LOW         (1 << 0)
/** GPIO pin is active (has logical value '1') in high state. */
#define GPIO_ACTIVE_HIGH        (0 << 0)

Above definitions are in include/dt-bindings/gpio/gpio.h, why they both have logical value '1'? it's for input or output?

**Comment by gabopushups** [comment]
Not an expert, but this is how I tell them apart. In the app code, there are two types, for the same device `dev`. ( "ret" is just a variable to check whether the operation was successful or not)
For an input:
`ret = gpio_pin_configure(dev, PIN, FLAGS);`

For an output:
`ret = gpio_pin_configure(dev, PIN, GPIO_OUTPUT_ACTIVE | FLAGS);`

**Comment by yashi** [comment, marked as answer]
"Logical value '1'" means that the GPIO is in active state as in comment string.

From https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html
> If a pin was configured as Active Low, physical level low will be considered as logical level 1 (an active state), physical level high will be considered as logical level 0 (an inactive state).

Let's say that a GPIO pin is connected to a chip's reset signal.  For some chips, the signal must be low to make the chip in reset state. For other chips, the signal must be high.  This depends on chips you are using.

But from the software point of view, we want to just say "Reset the chip" or "Activate the reset pin".  So, we, as in software, just want to consider the logical state instead of the hardware dependent physical signal level.  Both `GPIO_ACTIVE_LOW` and `GPIO_ACTIVE_HIGH` are for the active state, thus they both mean "logical 1".

And this is nothing to do with GPIO input / output.

**Reply by zzzh** [reply]
Thanks for the reply, it's really helpful.

Though, as you described, seems it's for output only.

Imagine you want to get an input to activate something,  it should be 0 or 1 to work? 

I think these names are confusing, it should include an in/out section.

**Reply by yashi** [reply]
> Though, as you described, seems it's for output only.

Apologize if I gave you a different impression.  But `GPIO_ACTIVE_LOW` and `GPIO_ACTIVE_HIGH` can be set to input pins as well.  It's orthogonal to the direction of the pins.

`gpio_pin_get()` is there if you want to use _logical_ levels as documented: https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html#c.gpio_pin_get
> Get logical level of an input pin taking into account GPIO_ACTIVE_LOW flag.

There is `gpio_pin_get_raw()` if you want to get the _physical_ level value.

> to get an input to activate something, it should be 0 or 1 to work?

It all depends on how you want it.  Do you want to activate it when the signal is physically low, or physically high?  If you have tones of signals with both active high and active low, just set `GPIO_ACTIVE_LOW` or `GPIO_ACTIVE_HIGH` when you `gpio_pin_configure()` and forget about it.  When you check the value, you don't have to worry about which pins you are handling, but just do
```c
if (gpio_pin_get(...)) {
   /* signal is active now, do something */
} else {
  /* signal is deactivated, do something else */
}
```

Again, if you don't want to use logical levels, just use physical raw levels.
```c
if (gpio_pin_get_raw(...)) {
   /* for this signal, high means deactivated, do something else */
} else {
   /* OK this is active, do something */
}
```

**Reply by zzzh** [reply]
Thank you so much for the reply, it's clear now.
